### ISSUE TITLE
gatk-haplotype-joint: slow GenomicsDBImport with large sample size

### ISSUE THREAD

**Issue by tkoomar** [bcbio/bcbio-nextgen#2852]
**tl;dr** I have a large WGS cohort (~400 samples). When running GenomicsDBImport as part of GATK, I noticed that importing regions larger than ~20Mb took a very long time. Several *large regions (80Mb+) ran for weeks without completing*.  Insight on how to speed up this step would be appreciated. Below I'll described the three approaches I took so far:

**Set `joint_group_size`**
In my configuration YAML I set
```
- algorithm:
    joint_group_size: 20
```
I am still a bit fuzzy on if this should have created 20 groups or groups of 20, but it had no apparent effect on the GenomicsDBImport calls. 

**Pass `--batch-size` argument to GATK**
I may have simply specified this wrong in my config YAML, but to try and force GATK to import in batches I added:
```
- algorithm:
  resources:
    gatk-haplotype-joint:
        options: ["--batch-size 50"]
```
This also had no apparent effect on the GenomicsDBImport calls. 

**Reduce Parallel Analysis Block Size**
Taking inspiration from #1097, I tried reducing `nomap_split_size` all the way down to 1 and increasing `nomap_split_targets` up to 3000, but there were still some 70Mb+ regions. My final though was to manually edit the `analysis_blocks.bed` to split large contiguous regions apart, but I wasn't sure if this was advisable.

**Comment by chapmanb** [comment]
Tanner;
Thanks for the detailed report and sorry about the issues with the runtimes. `GenomicsDBImport` is the central point of merging when all of the individual samples get put together, so is definitely a potential bottleneck. I'm not entirely sure about the slow runtimes but happy to offer suggestions:

- When you're having a long run, are you able to monitor the resource usage on the machine? It would be helpful to know if it's hitting memory issues with Java and swapping, or is just slow due to disk access issues. If the former, then `--batch-size` tweaking, as you tried, may be useful and we could try that next.

- Manually splitting the analysis blocks is a workaround to reduce sizes of individual parts that go into `GenomicsDBImport`. bcbio is conservative on splitting and won't split in regions of consistent coverage as it doesn't want to accidentally create a breakpoint at a SNP or indel, where it could lose the call due to splitting. That's pretty conservative so manually splitting where it refuses to is an option.

It would be great to understand the resource usage issues and offer workarounds in bcbio so this happens automatically, but hopefully either of these helps you get moving and finish your analysis. Thanks again.

**Comment by tkoomar** [comment]
Thanks Brad. I think specifying a batch-size will work. 

I tried manually running a `GenomicsDBImport` command on one of these large regions with a batch size of 5, and over the weekend it processed 45 out of 78 batches. This is the same region I previously let run for weeks without any output or update to the logs. [Here is a gist showing the command and resources actively used](https://gist.github.com/tkoomar/e5dda1ab88605b0c901defadf4bba8c7) (via `top`). One thing worth noting is that even though I  specified `--reader-threads 16`, the process appears to be using just 1. 

When run within the context of a bcbio run, nothing jumps out to me w/r/t memory swap. Here is a similar [gist showing resource usage within a bcbio run](https://gist.github.com/tkoomar/ac5bfd63b6b11a8b35317512fa7e5598). In this case I specified 3 cores per GATK process on a 56 thread machine, so there are 18 running simultaneously.  

I didn't specifically look disk I/O because this is being on a distributed nfs system that has never given me problems before, but I can try to dig into that if it would be helpful. 

Still being unfamiliar with GATK4's workflow, would testing this within the context of bcbio be as simple as hard-coding a `--batch-size ` argument in `gatkjoint.py` -- or or will this create some unexpected output for bcbio?
 https://github.com/bcbio/bcbio-nextgen/blob/ef886362376f36ba7ca6bb2ab27c3d377a300a69/bcbio/variation/gatkjoint.py#L53-L56

**Comment by chapmanb** [comment]
Tanner;
Thanks much for testing this out and glad to hear `--batch-size` resolves the issue. The latest development version automatically tweaks to use this on larger import sizes so hopefully it will just work if you update and re-start. Let us know if it still has issues with this approach and we can tweak further. Thanks again for reporting this and helping to debug.

**Comment by amizeranschi** [comment]
Below are the resource requests. Note: the cluster node `haswell-wn42.grid.pub.ro` is where the main bcbio process is executed, while the other nodes are running the `bcbio-c` and `bcbio-e` jobs.

```
$ grep "Resource requests:" /export/home/ncit/external/a.mizeranschi/automated-VC-test/testingVC/work/log/bcbio-nextgen-debug.log | head -n 10
[2020-01-22T13:20Z] haswell-wn42.grid.pub.ro: Resource requests: ; memory: 1.00; cores: 1
[2020-01-22T13:21Z] haswell-wn42.grid.pub.ro: Resource requests: bwa, sambamba, samtools; memory: 3.00, 3.00, 3.00; cores: 4, 4, 4
[2020-01-22T13:24Z] haswell-wn31.grid.pub.ro: Resource requests: ; memory: 1.00; cores: 1
[2020-01-22T13:24Z] haswell-wn31.grid.pub.ro: Resource requests: ; memory: 1.00; cores: 1
[2020-01-22T13:24Z] haswell-wn33.grid.pub.ro: Resource requests: ; memory: 1.00; cores: 1
[2020-01-22T13:24Z] haswell-wn31.grid.pub.ro: Resource requests: ; memory: 1.00; cores: 1
[2020-01-22T13:24Z] haswell-wn33.grid.pub.ro: Resource requests: ; memory: 1.00; cores: 1
[2020-01-22T13:24Z] haswell-wn38.grid.pub.ro: Resource requests: ; memory: 1.00; cores: 1
[2020-01-22T13:24Z] haswell-wn38.grid.pub.ro: Resource requests: ; memory: 1.00; cores: 1
[2020-01-22T13:24Z] haswell-wn33.grid.pub.ro: Resource requests: ; memory: 1.00; cores: 1

$ grep "Resource requests:" /export/home/ncit/external/a.mizeranschi/automated-VC-test/testingVC/work/log/bcbio-nextgen-debug.log | tail -n 10
[2020-01-22T13:24Z] haswell-wn39.grid.pub.ro: Resource requests: ; memory: 1.00; cores: 1
[2020-01-22T13:24Z] haswell-wn35.grid.pub.ro: Resource requests: ; memory: 1.00; cores: 1
[2020-01-22T13:24Z] haswell-wn32.grid.pub.ro: Resource requests: ; memory: 1.00; cores: 1
[2020-01-22T13:24Z] haswell-wn34.grid.pub.ro: Resource requests: ; memory: 1.00; cores: 1
[2020-01-22T13:24Z] haswell-wn40.grid.pub.ro: Resource requests: ; memory: 1.00; cores: 1
[2020-01-22T13:24Z] haswell-wn30.grid.pub.ro: Resource requests: ; memory: 1.00; cores: 1
[2020-01-22T13:24Z] haswell-wn38.grid.pub.ro: Resource requests: ; memory: 1.00; cores: 1
[2020-01-22T13:24Z] haswell-wn30.grid.pub.ro: Resource requests: ; memory: 1.00; cores: 1
[2020-01-22T13:24Z] haswell-wn34.grid.pub.ro: Resource requests: ; memory: 1.00; cores: 1
[2020-01-22T13:45Z] haswell-wn42.grid.pub.ro: Resource requests: gatk, gatk-haplotype, picard; memory: 1.00; cores: 1, 1, 1

```
